applications/convection_diffusion_application/custom_response_functions/local_temperature_average_response_function.h

```diff
@@ -0,0 +1,246 @@
+                                     const Matrix& rSensitivityMatrix,
+                                     Vector& rSensitivityGradient,
+                                     const ProcessInfo& rProcessInfo) override
+    {
+        if (rSensitivityGradient.size() != rSensitivityMatrix.size1())
+            rSensitivityGradient.resize(rSensitivityMatrix.size1(), false);
+        noalias(rSensitivityGradient) = ZeroVector(rSensitivityMatrix.size1());
+    }
+
+    void CalculatePartialSensitivity(Condition& rAdjointElement,
+                                     const Variable<array_1d<double, 3>>& rVariable,
+                                     const Matrix& rSensitivityMatrix,
+                                     Vector& rSensitivityGradient,
+                                     const ProcessInfo& rProcessInfo) override
+    {
+        if (rSensitivityGradient.size() != rSensitivityMatrix.size1())
+            rSensitivityGradient.resize(rSensitivityMatrix.size1(), false);
+        noalias(rSensitivityGradient) = ZeroVector(rSensitivityMatrix.size1());
+    }
+
+    double CalculateValue(ModelPart& rModelPart) override
+    {
+        KRATOS_TRY;
+        KRATOS_ERROR
+            << "PointTemperature::CalculateValue(ModelPart& rModelPart) is not implemented!!!\n";
+        KRATOS_CATCH("");
+    }
+
+    ///@}
+
+protected:
+    ///@name Protected member Variables
+    ///@{
+
+    ///@}
+    ///@name Protected Operators
+    ///@{
+
+    ///@}
+    ///@name Protected Operations
+    ///@{
+
+    ///@}
+
+private:
+    ///@name Member Variables
+    ///@{
+
+    ModelPart& mrModelPart;
+	int mNumNodes;
+
+    ///@}
+    ///@name Private Operators
+    ///@{
+
+    void ComputePointTemperatureSensitivityContribution(
+        const Matrix& rDerivativesOfResidual,
+        const Element::NodesArrayType& rNodes,
+        Vector& rLocalSensitivityContribution) const
+    {
+        if (rLocalSensitivityContribution.size() != rDerivativesOfResidual.size1())
+            rLocalSensitivityContribution.resize(rDerivativesOfResidual.size1(), false);
+
+        noalias(rLocalSensitivityContribution) = ZeroVector(rLocalSensitivityContribution.size());
+
+        const unsigned int num_nodes = rNodes.size();
+
+        for (unsigned int i = 0; i < num_nodes; i++)
+        {
+            if (rNodes[i].Is(STRUCTURE))
+            {
+                double factor = 1.0 / (rNodes[i].GetValue(NUMBER_OF_NEIGHBOUR_ELEMENTS)*mNumNodes);
+                rLocalSensitivityContribution[i] = factor;
+            }
+        }
+    }
+
+    ModelPart& GetTargetModelPart(ModelPart& rModelPart, const std::string& rTargetModelPartName)
+    {
+        if (rModelPart.Name() == rTargetModelPartName)
+		{
+            return rModelPart;
+		}
+		else if (rModelPart.HasSubModelPart(rTargetModelPartName))
+		{
+			return rModelPart.GetSubModelPart(rTargetModelPartName);
+		}
+		else
+		{
+			KRATOS_ERROR << "Unknown ModelPart " << rTargetModelPartName << "." << std::endl;
```

Comment:
Maybe there should still be a return statement here (after the error).

Reply:
I added a default return in case execution reaches the end of the function, although it should have already thrown in that case.

Reply:
Ya, I still think it is good since it eliminates possible compiler warnings and lets the reader know, this wasn't simply forgotten.